### ISSUE TITLE
Run handlers with URI

### ISSUE THREAD

**Issue by reygoch** [haskell-servant/servant#1147]
Is it possible to run servant handlers with just URI? Meaning, can I run servant application without starting the e.g. wai server?

**Comment by alpmestan** [comment]
Haha, nice one @phadej :-)

@reygoch Something a bit like this should be possible, in theory yes. This is the kind of feature that motivates me for thinking a bit more about extracting all the HTTP-server independent routing machinery into its own package, with a few utilities to inspect/debug/evaluate routers (which is what your question is about). This might be pushing me enough to get started....

As an aside, it's indeed not clear what the best return type should be. Not even mentionning that just the URI isn't enough, a servant API type may refer to cookies (for auth), any header, the response body... So you'd probably be passing something that looks a lot like a `Request` type, one way or another, wouldn't you? The obvious decision for the return type then could be just a `Response` type. We would have to use TH or some dark magic to do anything smarter than that. Or we could indeed try leaving it for the user to decide, I'm not sure what solution is the best.

**Comment by reygoch** [comment]
@alpmestan yes, it occurred to me before I went to sleep yesterday that input should be a `Request` 😄, although I don't think I agree with the `Response` part necessarily since I'd just want to get a result from the handler.

Maybe there could be 4 main utility functions for this purpose. Where functions that take `URI` are just a wrapper around the ones which take `Request` mainly for the simplicity of running simpler requests which are only `URI` dependent.

```haskell
runServerless :: Server api -> Request -> Either ServerError a
runServerlessURI :: Server api -> URI -> Either ServerError a
runServerless' :: Server api -> Request -> Response
runServerlessURI :: Server api -> URI -> Response
```

As for the motivation (if you need some more), I've been playing with Miso library and have concluded that this feature would be very useful for initializing client side model in "isomorphic" apps.

Right now, if the user directly accesses certain `URI` e.g. `user/{id}` after Miso app boots up I need to fetch the app model / state from the server because I can't parse out the user `id` from the `URI` and execute my `Action` to load the user data which would feel more natural.

Another thing is that I have to essentially maintain two endpoints for `user/{id}` route on the server side. One that renders the model into the view and returns `HTML` view and another one which returns my `Model` as `JSON` (this is what client would call to initialize his model in case user directly accessed a certain `URI`).
